torchaudio/prototype/tacotron2.py

```diff
@@ -0,0 +1,851 @@
+                                attention_location_n_filters,
+                                attention_location_kernel_size,
+                                attention_rnn_dim, decoder_rnn_dim,
+                                prenet_dim, max_decoder_steps,
+                                gate_threshold, p_attention_dropout,
+                                p_decoder_dropout,
+                                not decoder_no_early_stopping)
+        self.postnet = _Postnet(n_mels, postnet_embedding_dim,
+                                postnet_kernel_size,
+                                postnet_n_convolutions)
+
+    def _parse_output(self,
+                      outputs: Tuple[Tensor, Tensor, Tensor, Tensor],
+                      output_lengths: Tensor) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
+
+        if self.mask_padding and output_lengths is not None:
+            mask = get_mask_from_lengths(output_lengths)
+            mask = mask.expand(self.n_mels, mask.size(0), mask.size(1))
+            mask = mask.permute(1, 0, 2)
+
+            outputs[0].masked_fill_(mask, 0.0)
+            outputs[1].masked_fill_(mask, 0.0)
+            outputs[2].masked_fill_(mask[:, 0, :], 1e3)  # gate energies
+
+        return outputs
+
+    def forward(self,
+                text: Tensor,
+                text_lengths: Tensor,
+                mel_specgram: Tensor,
+                mel_specgram_lengths: Tensor) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
+        r"""Pass the input through the Tacotron2 model. This is in teacher
+        forcing mode, which is generally used for training.
+
+        The input ``text`` should be padded with zeros to length text_lengths.max().
+        The input ``mel_specgram`` should be padded with zeros to length mel_specgram_lengths.max().
+
+        Args:
+            text (Tensor): The input text to Tacotron2 with shape (n_batch, text_lengths.max()).
+            text_lengths (Tensor): The length of each text with shape (n_batch).
+            mel_specgram (Tensor): The target mel spectrogram with shape (n_batch, n_mels, mel_specgram_lengths.max()).
+            mel_specgram_lengths (Tensor): The length of each mel spectrogram with shape (n_batch).
+
+        Returns:
+            mel_specgram (Tensor): Mel spectrogram before Postnet
+                with shape (n_batch, n_mels, mel_specgram_lengths.max()).
+            mel_specgram_postnet (Tensor): Mel spectrogram after Postnet
+                with shape (n_batch, n_mels, mel_specgram_lengths.max()).
+            stop_token (Tensor): The output for stop token at each time step
+                with shape (n_batch, mel_specgram_lengths.max()).
+            alignment (Tensor): Sequence of attention weights from the decoder.
+                with shape (n_batch, mel_specgram_lengths.max(), text_lengths.max()).
+        """
+
+        text_lengths, mel_specgram_lengths = text_lengths.data, mel_specgram_lengths.data
```

Comment:
why is `.data` needed here? in general, we should never need to call `.data`.

Reply:
addressed in the new [commit](https://github.com/pytorch/audio/pull/1621/commits/5fccd01181b8a5d5aa3e33f0f0102c27bd154223).
Thanks.